modules/dnn/src/model.cpp

```diff
@@ -0,0 +1,212 @@
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+#include <algorithm>
+#include <iostream>
+#include <utility>
+#include <iterator>
+
+#include <opencv2/imgproc.hpp>
+
+namespace cv {
+namespace dnn {
+
+struct Model::Impl {
+    Size   size;
+    Scalar mean;
+    float  scale = 1.0;
+    bool   swapRB = false;
+    bool   crop = false;
+    Mat    blob;
+    std::vector<String> outNames;
+
+    void predict(Net* net, const Mat& frame, std::vector<Mat> outs) {
+        blobFromImage(frame, blob, 1.0, size, Scalar(), swapRB, crop, CV_8U);
+        net->setInput(blob, "", scale, mean);
+
+        outNames = net->getUnconnectedOutLayersNames();
+        net->forward(outs, outNames);
+    }
+};
+
+Model::Model(const std::string& model, const std::string& config)
+    : Net(readNet(model, config)), impl_(new Impl) {};
+
+Model::Model(const Net& network) : Net(network), impl_(new Impl) {};
+
+Model& Model::setInputSize(const Size& size) {
+    impl_->size = size;
+    return *this;
+}
+
+Model& Model::setInputSize(int width, int height) {
+    impl_->size = Size(width, height);
+    return *this;
+}
+
+Model& Model::setInputMean(const Scalar& mean) {
+    impl_->mean = mean;
+    return *this;
+}
+
+Model& Model::setInputScale(float scale) {
+    impl_->scale = scale;
+    return *this;
+}
+
+Model& Model::setInputCrop(bool crop) {
+    impl_->crop = crop;
+    return *this;
+}
+
+Model& Model::setInputSwapRB(bool swapRB) {
+    impl_->swapRB = swapRB;
+    return *this;
+}
+
+void Model::predict(InputArray frame, OutputArray outs) {
+    impl_->predict(this, frame.getMat(), outs.getMat());
+}
+
+ClassificationModel::ClassificationModel(const std::string& model, const std::string& config)
+    : Model(model, config) {};
+
+ClassificationModel::ClassificationModel(const Net& network) : Model(network) {};
+
+std::pair<int, float> ClassificationModel::classify(InputArray frame) {
+    blobFromImage(frame, impl_->blob, 1.0, impl_->size, Scalar(), impl_->swapRB, impl_->crop, CV_8U);
+    setInput(impl_->blob, "", impl_->scale, impl_->mean);
+    Mat out = forward();
+
+    double conf;
+    cv::Point maxLoc;
+    minMaxLoc(out, nullptr, &conf, nullptr, &maxLoc);
+
+    return {maxLoc.x, static_cast<float>(conf)};
+}
+
+DetectionModel::DetectionModel(const std::string& model, const std::string& config)
+    : Model(model, config) {};
+
+DetectionModel::DetectionModel(const Net& network) : Model(network) {};
+
+void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
+                   CV_OUT std::vector<float>& confidences, CV_OUT std::vector<Rect2d>& boxes,
+                   float confThreshold, float nmsThreshold, bool absoluteCoords)
+{
+    int frameWidth  = frame.cols();
+    int frameHeight = frame.rows();
+
+    if (impl_->size.width <= 0) impl_->size.width = frame.cols();
+    if (impl_->size.height <= 0) impl_->size.height = frame.rows();
```

Comment:
We need to think about three options about input sizes:

1) blob size is specified
1) blob size == frame size
3) blob size is from model (not implemented yet).

Reply:
Need to add assert?

Reply:
Not actually. I mean that it could be useful to have a feature with automatic input blob size detection from the model. However it won't work for all the models because some formats have no information about input sizes.

So I agree with you. Let's add an assertion if `width` or `height` weren't specified.

Reply:
Maybe add 
```
if (size.width <= 0) size.width = frame.cols;
if (size.height <= 0) size.height = frame.rows;
```
to Impl::predict(Net& net, const Mat& frame, std::vector<Mat>& outs)?